src/powerpwn/copilot_studio/modules/deep_scan.py

```diff
@@ -471,19 +475,110 @@ def print_brand(tenant: str, timeout: int = 10):
         print(f"No brand found for tenant {tenant}")
 
 
+def run_pup_commands(existing_bots):
+    if os.name == "nt":  # Windows
+        pup_path = get_project_file_path("tools/pup_is_webchat_live", "is_chat_live_windows.js")
+    else:
+        pup_path = get_project_file_path("tools/pup_is_webchat_live", "is_chat_live.js")
+    # Construct the command to run the Node.js script with xargs
+    open_bots_path = get_project_file_path("final_results/", "chat_exists_output.txt")
+    # Empty the file
+    with open(open_bots_path, "w") as _:
+        pass
+    for bot_url in existing_bots:
+        try:
+            # Construct the shell command
+            command = f"node {pup_path} {bot_url}"
+            logging.debug(f"Running command: {command}")
+            # TODO: Verify and improve guardrails for using subprocess + replace shell=True
+            # Run the command
+            subprocess.run(command, shell=True, check=True)  # nosec
+        except subprocess.CalledProcessError as e:
+            logging.error(f"Error occurred while running puppeteer: {e}")
+
+    if os.path.exists(open_bots_path):
+        return sort_unique_values_in_file(open_bots_path)
+    return []
+
+
+def get_bot_name_from_url(bot_url):
+    re_match = re.search(r"/bots/(.*?)/canvas", bot_url)
+    if re_match:
+        return re_match.group(1)
+    return ""
+
+
 class DeepScan:
     """
     A class that is responsible for the CPS deep scan
     """
 
     def __init__(self, args):
         self.args = args
+        self.default_env_found = False
+        self.default_solution_prefix_found = False
+        self.existing_bots = []
+        self.open_bots = []
         self.run()
 
-    def run(self):
+    def dump_results(self):
```

Comment:
Let's please add _docstrings_ with explicit types & documentation in the passed parameters for best practice (see existing functions).